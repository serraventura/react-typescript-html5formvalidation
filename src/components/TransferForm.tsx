import React, { useState, useContext } from "react";
import { TextField, Button, Header3 } from "@zopauk/react-components";
import { ThemeContext } from "./ThemeContext";
import { Form, Lead } from "./styled";
import { ITransferForm, IErrorFormValidation } from "./types";

const initialErrorFormValidation: IErrorFormValidation = {
  name: false,
  email: false,
  amount: false
};

const TransferForm: React.FC<ITransferForm> = ({ onSubmit }) => {
  const context = useContext(ThemeContext);
  const [name, setName] = useState<string>("");
  const [email, setEmail] = useState<string>("");
  const [amount, setAmount] = useState<number>(0);
  const [error, setError] = useState<IErrorFormValidation>(
    initialErrorFormValidation
  );

  return (
    <Form
      data-test="transfer-form" 
      color={context.base.white}
      onSubmit={e => {
        onSubmit({ name, email, amount });
        e.preventDefault();
        setName("");
        setEmail("");
        setAmount(0);
        setError(initialErrorFormValidation);
        return;
      }}
    >
      <Header3>Enter details</Header3>
      <Lead>
        <TextField
          label="Name"
          errorMessage={error.name ? "This field is required" : ""}
          inputProps={{
            name: "name",
            required: true,
            type: "text",
            value: name,
            onChange: e => {
              const { target } = e;
              setName(target.value);
              setError({ ...error, name: !target.checkValidity() });
            }
          }}
        />
      </Lead>
      <Lead>
        <TextField
          label="E-mail"
          errorMessage={
            error.email ? "Provide a valid e-mail. This field is required" : ""
          }
          inputProps={{
            name: "email",
            required: true,
            type: "email",
            value: email,
            onChange: e => {
              const { target } = e;
              setEmail(target.value);
              setError({ ...error, email: target.validity.typeMismatch });
            }
          }}
        />
      </Lead>
      <Lead>
        <TextField
          label="Amount"
          errorMessage={
            error.amount ? "Invalid value. This field is required" : ""
          }
          inputProps={{
            name: "amount",
            required: true,
            type: "number",
            step: "any",
            value: amount,
            onChange: e => {
              const { target } = e;
              setAmount(+target.value);
              setError({ ...error, amount: !target.checkValidity() });
            }
          }}
          prefix={"Â£"}
        />
      </Lead>

      <Button type="submit">Send</Button>
    </Form>
  );
};

export default TransferForm;
