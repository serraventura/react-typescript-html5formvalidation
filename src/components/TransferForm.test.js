import React from "react";
import { shallow, mount, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import TransferForm from "./TransferForm";

configure({ adapter: new Adapter() });

let onSubmit;

describe("<TransferForm />", () => {
  beforeEach(() => {
    onSubmit = jest.fn();
  });

  it("should render without crashing", () => {
    const wrapper = shallow(<TransferForm />);
    expect(wrapper).toBeTruthy();
  });

  it("should match snapshot", () => {
    const wrapper = mount(<TransferForm />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should handle onSubmit prop", () => {
    const wrapper = shallow(<TransferForm onSubmit={onSubmit} />);

    wrapper
      .find('*[data-test="transfer-form"]')
      .first()
      .simulate("submit", { preventDefault: () => {} });

    expect(onSubmit).toHaveBeenCalled();
    expect(onSubmit).toHaveBeenCalledWith({ amount: 0, email: "", name: "" });
  });
});
