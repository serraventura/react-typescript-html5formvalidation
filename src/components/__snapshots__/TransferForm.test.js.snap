// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TransferForm /> should match snapshot 1`] = `ReactWrapper {}`;
